server/connectors/cloudbees_connector/oc_client.py

```
@@ -0,0 +1,279 @@
+
+    def _validate_controller_url(self, controller_url: str) -> None:
+        """Validate that a controller URL belongs to the same registrable domain as the OC base URL.
+
+        Allows sibling subdomains (e.g., controller.company.com when OC is oc.company.com)
+        but blocks different domains (e.g., evil.attacker.com).
+        """
+        oc_host = urlparse(self.base_url).hostname or ""
+        ctrl_host = urlparse(controller_url).hostname or ""
+
+        if not ctrl_host:
+            raise ValueError(f"Invalid controller URL: {controller_url}")
```

Ruff (0.15.15)

[warning] Avoid specifying long messages outside the exception class

(TRY003)